woocommerce/woocommerce
# Fix failing smoke tests when a new release candidate is published

### All Submissions:

### Changes proposed in this Pull Request:

The first step in publishing is to use `Woorelease` which creates a release on GitHub. Next, the `build-release.yml` workflow gets triggered, which packages and uploads a `woocommerce.zip` asset for that release. 

The reason why the `smoke-test-release.yml` workflow [fails](https://github.com/woocommerce/woocommerce/runs/5054386785?check_suite_focus=true) is because it gets triggered on release publish events, and in this case, it wouldn't have an asset to download until after the `build-release.yml` workflow is complete.

The fix proposed is to have the `build-release.yml` workflow trigger the `smoke-test-release.yml` after it has uploaded the new asset. That way `smoke-test-release.yml` won't run into the issue of having a release with no asset attached to the issue.

Things to note:
- This workflow adds `E2E_WORKFLOW_GH_TOKEN` to our secrets. A reminder has been set up to prompt the team to update the token a day before it expires. 
- The `workflow_dispatch` changes won't be picked up by GitHub until this branch is merged(workflows with that event trigger must be on the default branch to be recognized by GitHub)
- Since this depends on a RC to be published, it may make sense to have a test PR after merging to ensure this implementation works as expected.

Closes #31833.

### How to test the changes in this Pull Request:
I've created a public repository at https://github.com/MrJnrman/woo-workflow-test to test the core of these changes. Its basically a copy-paste of the logic to trigger the `build-release.yml` on the publishing of a release, which subsequently triggers a workflow that runs the `Download WooCommerce release zip` workflow step.

1. Clone or fork the repository
2. [Generate](https://github.com/settings/tokens) an access token with permissions for `workflows`
3. Create a project secret called `CUSTOM_GH_TOKEN` with its value set to the generated access token
4. Create the following releases and observe the workflows from the `Actions`:
   - A pre-release with an asset
   - A pre-release without an asset
   - A release with an asset
   - A release without an asset

**Note: Do not upload a zip with the name `woocommerce.zip` this will break the workflow**
